libclamav/readdb.c

```diff
@@ -1191,37 +1191,46 @@ static int cli_chkign(const struct cli_matcher *ignored, const char *signame, co
 
 static int cli_chkpua(const char *signame, const char *pua_cats, unsigned int options)
 {
-    char cat[32], *pt;
+    char cat[32], *cat_pt, *pt1, *pt2, *endsig;
     const char *sig;
     int ret;
 
+    cli_dbgmsg("cli_chkpua: Checking signature [%s]\n", signame);
+
     if (strncmp(signame, "PUA.", 4)) {
         cli_dbgmsg("Skipping signature %s - no PUA prefix\n", signame);
         return 1;
     }
     sig = signame + 3;
-    if (!(pt = strchr(sig + 1, '.'))) {
+    if (!(pt1 = strchr(sig + 1, '.'))) {
         cli_dbgmsg("Skipping signature %s - bad syntax\n", signame);
         return 1;
     }
-
-    if ((unsigned int)(pt - sig + 2) > sizeof(cat)) {
-        cli_dbgmsg("Skipping signature %s - too long category name\n", signame);
+    if ( (pt2 = strrchr(sig + 1, '.')) != pt1 ) {
+        cli_dbgmsg("Signature has at least three dots [%s]\n", signame);
+    }
+    if ((unsigned int)(pt1 - sig + 2) > sizeof(cat)) {
+        cli_dbgmsg("Skipping signature %s - too long category name, length approaching %d characters\n", signame, (unsigned int)(pt1 - sig + 2) );
+        return 1;
+    }
+    if ((unsigned int)(pt2 - sig + 2) > sizeof(cat)) {
+        cli_dbgmsg("Skipping signature %s - too long category name, length approaching %d characters\n", signame, (unsigned int)(pt2 - sig + 2));
         return 1;
     }
 
-    strncpy(cat, sig, pt - signame + 1);
-    cat[pt - sig + 1] = 0;
-    pt                = strstr(pua_cats, cat);
-
+    endsig = strrchr(sig, '.');
+    strncpy(cat, sig, strlen(sig) - strlen(endsig) + 1);
+    cat[strlen(sig) - strlen(endsig) + 1] = 0;
+    cat_pt                = strstr(pua_cats, cat);
```

Comment:
I have played with this PR a little, and I am finding that if I use just the category, then a signature is not excluded.  For example.

If I have signatures PUA.Cat.subcat.sig, and my command is --exclude-pua=Cat, they are not excluded.  It looks like the arguments for strstr are backwards.  pua_cats is actually the 'needle', and cat is the 'haystack' (It is currently incorrect in main).  

I have tried switching this call to 'cat_pt                = strstr(cat, pua_cats);', and it excludes as expected.

Reply:
I will flip these two params and update the PR

Reply:
PR updated